design/src/rfcs/rfc0024_supporting_select_object_content_scan_range.md

```diff
@@ -0,0 +1,214 @@
+
+How to actually implement this RFC
+----------------------------------
+
+In order to implement this feature, we would extend the current S3 customization to prevent `ScanRange` from being
+generated normally and would instead insert inlineable code into `aws_sdk_s3::models` that provided a handwritten
+version of `ScanRange` with special behavior. We would generate a serialization function as well that handles
+defaults and "unset" fields as defined in the `ScanRange` docs.
+
+By generating an enum instead of a struct, we could allow users to be explicit about the behavior they expect.
+
+```rust
+#[non_exhaustive]
+#[derive(Clone, PartialEq)]
+pub enum ScanRange {
+    StartToEnd {
+        start: i64,
+        end: i64,
+    },
+    StartToEndOfObject {
+        start: i64,
+    },
+    LastNBytesOfObject {
+      n_bytes: i64
+    }
+}
+
+impl ScanRange {
+    pub fn start_to_end(start: i64, end: i64) -> Self {
+        ScanRange::StartToEnd { start, end }
+    }
+
+    pub fn start_to_end_of_object(start: i64) -> Self {
+        ScanRange::StartToEndOfObject { start }
+    }
+
+    pub fn last_n_bytes_of_object(n_bytes: i64) -> Self {
+      ScanRange::LastNBytesOfObject { n_bytes }
+    }
+
+    pub fn start(&self) -> Option<i64> {
+        match self {
+            ScanRange::StartToEnd { start, .. } => Some(start),
+            ScanRange::StartToEndOfObject { start } => Some(start),
+            ScanRange::LastNBytesOfObject { .. } => None,
+        }
+    }
+
+    pub fn end(&self) -> Option<i64> {
+        match self {
+            ScanRange::StartToEnd { end, .. } => Some(end),
+            ScanRange::StartToEndOfObject { .. } => None,
+            ScanRange::LastNBytesOfObject { n_bytes } => Some(n_bytes),
+        }
+    }
+}
+```
+
+To ensure that this handwritten struct doesn't deviate from the S3 model, we'd create a codegen test that would verify
+compatibility of the two models.
+
+Why this RFC should be rejected
+-------------------------------
+
+There are several reasons to reject this RFC:
+
+- Writing customizations to make up for the modeling mistakes of services is not a scalable strategy. Any customizations
+  we write won't necessarily be implemented by other SDKs, causing a fragmented and annoying user experience.
+- The proposed behavior is confusing and difficult to explain.
+- Differentiating between "set", "default", and "unset" struct fields is not currently supported by the SDK. Adding it
+  would take a few weeks of effort. _(We do need to eventually support this anyway for IDL v2.)_
+- [Only one user][user-submitted issue] has asked for this feature at the time of writing this RFC.
+
+### What we should do instead
+
+We have already committed to supporting [IDLv2] and implementing that would allow users to set `start=0` or unset it.
+Although IDLv2 support would NOT support the custom `end` value behavior, we'd be halfway there. Any future S3
```

Comment:
This is a little confusing because it seemed like the handwritten `ScanRange` implementation was to be applied to the IDLv2 implementation. But this sentence makes it seem like the `ScanRange` proposal is actually for an IDLv1 implementation?